src/libstd/sys/unix/fs.rs

```diff
@@ -51,6 +51,7 @@ pub struct FileAttr {
 pub struct ReadDir {
     dirp: Dir,
     root: Arc<PathBuf>,
+    end_of_stream: bool,
```

Comment:
I think it might be better to just wrap `Dir` into an `Option` (and mark the pointer within `Dir` as `NonZero`) as that would both be more semantically accurate, more efficient space wise and more future-proof in terms of preventing mistakes.